pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/ModelInterpreter.groovy

```diff
@@ -180,49 +181,55 @@ class ModelInterpreter implements Serializable {
             def thisError = null
             script.stage(thisStage.name) {
                 try {
-                    if (firstError != null) {
-                        Utils.logToTaskListener("Stage '${thisStage.name}' skipped due to earlier failure(s)")
-                        Utils.markStageSkippedForFailure(thisStage.name)
-                        isSkipped = true
-                        if (thisStage.parallel != null) {
-                            script.parallel(getParallelStages(root, parentAgent, thisStage, firstError, parentStage, true, false, false))
-                        }
-                    } else if (skipUnstable(root.options)) {
-                        Utils.logToTaskListener("Stage '${thisStage.name}' skipped due to earlier stage(s) marking the build as unstable")
-                        Utils.markStageSkippedForUnstable(thisStage.name)
-                        isSkipped = true
-                        if (thisStage.parallel != null) {
-                            script.parallel(getParallelStages(root, parentAgent, thisStage, firstError, parentStage, false, true, false))
-                        }
-                    } else {
-                        if (thisStage.parallel != null) {
-                            if (evaluateWhen(thisStage.when)) {
-                                withCredentialsBlock(thisStage.environment) {
-                                    withEnvBlock(thisStage.getEnvVars(script)) {
-                                        script.parallel(getParallelStages(root, parentAgent, thisStage, firstError, parentStage, false, false, false))
+                    inWrappers(thisStage.options?.wrappers) {
```

Comment:
Would be nice if we could have a two stage wrappers. One here and one inside the agent (for those wrappers requiring a workspace) Then people could use file credentials etc.

Reply:
And should we really not check for `firstError` etc. before entering the wrappers? they could produce quite a lot of computation for no use since the stage will be "skipped" anyway.

Reply:
Lemme think on this.

Reply:
Yeah, agreed on the error case - tweaking.